Narrow types in event listeners inside conditional blocks

crisbeto opened the pull request:
Currently we rely on native `if` and `switch` blocks to do the type narrowing for us in the new control flow syntax. This works for most cases, except for event listeners. Since event listener expressions are inside callback functions, type narrowing doesn't apply to them anymore. These changes work around it by reconstructing an expression to re-narrow the type in such cases.

Fixes #52052.

crisbeto committed:
fix(compiler): narrow the type of expressions in event listeners inside if blocks

Since expressions in event listener are added inside of a callback, type narrowing won't apply to them anymore. These changes add the logic to create a guard expression that will re-narrow the expression in the callback.

Fixes #52052.

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1231,29 +1233,80 @@ class TcbIfOp extends TcbOp {
+      const expression = this.branchExpressions.get(branch) as ts.Expression | null;
+
+      // Skip over branches without an expression.
+      if (expression !== null) {
```
nit: compiler generally prefers early exits to reduce nesting

```suggestion
      if (expression === null) {
        continue;
      }
```

atscott commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1289,12 +1344,49 @@ class TcbSwitchOp extends TcbOp {
+          guard = expression;
+        } else {
+          guard = ts.factory.createBinaryExpression(
+              guard, ts.SyntaxKind.AmpersandAmpersandToken, expression);
```
Do we need `markIgnoreDiagnostic` somewhere here? similar to the reasoning from the other PR https://github.com/angular/angular/pull/52110/files#r1350466094

crisbeto committed:
fix(compiler): narrow the type of expressions in event listeners inside switch blocks

Since expressions in event listener are added inside of a callback, type narrowing won't apply to them anymore. These changes add the logic to create a guard expression that will re-narrow the expression in the callback.

Fixes #52052.

crisbeto commented:
Rebased and addressed the feedback.

crisbeto commented:
Caretaker note: please ignore the Windows failures. The failures are unrelated and the CI seems to have gotten stuck in the flaky state.
